### PR TITLE
Migration to torch distributions and scoringrules integration

### LINKED ISSUE #40: Migrate to Keras 3
Migrate library to use Keras 3 https://keras.io/keras_3/

### PR DESCRIPTION
**Pull request by MicheleCattaneo**
# Proposal

- Remove the dependency of TFP in order to move to Keras3 and rely on torch as backend and `torch.distributions` for the probabilistic layers.
- Rely on [scoringrules](https://github.com/frazane/scoringrules) for probabilistic loss functions.

## Current progress 

- [x] Update poetry dependencies.
- [x] scoringrules has been tested on normal and exponential distributions with crps.
- [ ] Update code to remove `tf` dependencies.
- [x] Update all models to the new structure
  - TCN and phyisical layers are temporarily not supported.
- [ ] Update tests to the new structure (uncomment one test at the time, run and fix)
    - [ ] Cleanup warnings
- [x] Create the equivalent of `DistributionLossWrapper` for sample losses.
- [ ] Ensure configuration files are compatible with the updates
- [ ] Optional torch CPU/GPU. See the poetry documentation on mutually exclusive "extras" [here](https://python-poetry.org/docs/dependency-specification/#exclusive-extras).
- [x] Add distributions:
  - [x] Censored normal (custom)
  - [x] Truncated normal (custom)
  - [x] Multivariate normal
  - [x] Gamma
  - [x] Beta
  - [x] LogNormal
- [ ] Add "pre-coocked" losses:
  - [x] WeightedCRPSEnergy
  - [ ] EnergyScore
- [ ] Update README.md (via notebook)
  - [ ] Add all the out-of-the-box supported distributions  (i think it's a nice "selling" point)
- [ ] Add test for SHAP (include as dev dependency)
- [x] Extend support to python 3.12 
- [x] Format with black
- [ ] ...

### Some implementation details: 

- Every distribution will implement `BaseParametricDistribution` 
- A probabilistic layer will implement `BaseDistributionLayer`, which contains an underlying distribution and is responsible for mapping its input into the correct number of parameters.
- `DistributionLossWrapper` allows to wrap a function from scoringrules such that the loss can be computed between a tensor of observations and a predicted distribution.
- There is a clear separation between models and layers.

Useful links:
- Table with common `tf` operations and their corresponding keras3 variant: [here](https://keras.io/guides/migrating_to_keras_3/#:~:text=in%20this%20guide.-,NumPy%20ops,-Keras%20implements%20the)

### CODE REVIEW

**Comment by MicheleCattaneo** [comment]
Serializing wrapped torch modules (via `keras.layers.TorchModuleWrapper`) seems to be problematic, as reported [here](https://github.com/keras-team/keras/issues/19226) and [here](https://github.com/keras-team/keras/issues/20860).
In case this is definitely a bug we can:
- Try to fix the bug ourselves and propose a PR
- Rely on `save_weights` and `load_weights` rather than `keras.models.save`

The first option would require us to wait for a fix. The second option would require to manually log models on mlflow as generic binary files, as `mlflow.keras.log_model` internally uses `keras.model.save`

**Comment by dnerini** [comment]
[keras 3.11](https://github.com/keras-team/keras/releases/tag/v3.11.0) has been released, it includes the fix for the serialization of wrapped torch modules (see https://github.com/keras-team/keras/issues/20860)!

File: mlpp_lib/probabilistic_layers.py [hunk 1]
```diff
@@ -1,1462 +1,1943 @@
...
-            "convert_to_tensor_fn": _serialize(self._convert_to_tensor_fn),
-            "validate_args": self._validate_args,
-        }
-        base_config = super(IndependentGamma, self).get_config()
-        return dict(list(base_config.items()) + list(config.items()))
-
+    def mean(self):
+        return self._distribution.mean
+        
+class UniveriateGaussianModule(BaseParametricDistributionModule):
```
Comment (dnerini) [inline mlpp_lib/probabilistic_layers.py]: typo
